File: apps/ai-service/app/schemas/ai.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator

LEGACY_POLISH_TONES = {
    "professional": "专业",
    "casual": "轻松自然",
    "technical": "技术严谨",
    "grammar": "严谨准确",
    "clarity": "清晰易懂",
    "style": "自然流畅",
    "all": "专业",
}


def _normalize_tone(value: object | None) -> str | None:
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None
    return LEGACY_POLISH_TONES.get(text.lower(), text)


class SummaryRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100000)
    maxLength: int = Field(default=200, ge=10, le=2000)
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False


class ExistingTagHint(BaseModel):
    """已有标签提示。前端在请求生成标签时携带,让 AI 优先复用而非新建。

    `name` 是标签人类可读名;`postCount` 是该标签关联的文章数,用于在 prompt 中
    按热度排序展示给模型,也用于前端 UI 渲染时的"热门度"徽标。
    """

    name: str = Field(..., min_length=1, max_length=64)
    postCount: int = Field(default=0, ge=0)


class TagMatch(BaseModel):
    """AI 命中的"现有标签"。`reason` 是模型可选给出的一句话匹配理由。"""

    name: str
    postCount: int = 0
    reason: Optional[str] = None


class TagsRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100000)
    maxTags: int = Field(default=5, ge=1, le=20)
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False
    # 现有标签库提示。可选;为空时退化为旧行为 (纯新建议)。
    # 上限 200 项是 prompt 体积 + token 成本的工程取舍:典型博客标签数 < 100,
    # 极端情况下也能容纳大型站点的高频段。前端按 postCount DESC 截断。
    existingTags: list[ExistingTagHint] = Field(default_factory=list, max_length=200)


class TitlesRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100000)
    maxTitles: int = Field(default=5, ge=1, le=10)
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False
    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="before")
    @classmethod
    def normalize_legacy_fields(cls, values):
        if not isinstance(values, dict):
            return values
        normalized = dict(values)
        if normalized.get("maxTitles") in (None, "") and normalized.get("count") not in (None, ""):
            normalized["maxTitles"] = normalized["count"]
        return normalized


class PolishRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100000)
    tone: Optional[str] = None
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False
    model_config = ConfigDict(populate_by_name=True)

    @model_validator(mode="before")
    @classmethod
    def normalize_legacy_fields(cls, values):
        if not isinstance(values, dict):
            return values
        normalized = dict(values)
        tone = _normalize_tone(normalized.get("tone"))
        if tone is None:
            tone = _normalize_tone(normalized.get("style"))
        if tone is None:
            tone = _normalize_tone(normalized.get("polishType"))
        if tone is not None:
            normalized["tone"] = tone
        return normalized


class OutlineRequest(BaseModel):
    topic: Optional[str] = Field(default=None, min_length=1)
    content: Optional[str] = None
    existingContent: Optional[str] = None
    depth: int = Field(default=2, ge=1, le=6)
    style: str = Field(default="professional")
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False

    @model_validator(mode="after")
    def validate_topic_or_content(self):
        if not self.topic and not self.content:
            raise ValueError("topic or content is required")
        return self


class SummaryData(BaseModel):
    summary: str
    characterCount: int
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None


class TagsData(BaseModel):
    """标签生成结果。

    - ``tags``:    扁平字符串数组,匹配 + 新建议合并 (保留旧契约,旧客户端不动)。
    - ``matches``: 命中现有标签的结构化列表 (含 ``postCount`` / ``reason``)。
    - ``suggestions``: 现有标签库未覆盖,需要新建的标签名。

    新客户端应优先消费 ``matches`` / ``suggestions``;``tags`` 仅作向后兼容。
    """

    tags: list[str]
    matches: list[TagMatch] = Field(default_factory=list)
    suggestions: list[str] = Field(default_factory=list)
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None


class TitlesData(BaseModel):
    titles: list[str]
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None


class PolishData(BaseModel):
    polishedContent: str = Field(alias="polishedContent")
    # Note: 历史上曾设计过 `changes: Optional[str]` 用于返回"变更说明"，但端点
    # 从未写入该字段，前端也无法区分"LLM 真返回了空" vs "该字段被放弃"，
    # 长期停留在接口文档中反而误导消费者。2026-04 的 AI 工具修复已将其移除。
    # 若未来需要 diff/变更说明，请通过独立的 `/api/v1/ai/polish/diff` 端点提供。
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None

    class Config:
        populate_by_name = True


class OutlineData(BaseModel):
    outline: str
    characterCount: int
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None


class TranslateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=100000)
    targetLanguage: str = Field(default="en")
    sourceLanguage: Optional[str] = None
    promptVersion: Optional[str] = None
    promptTemplate: Optional[str] = None
    modelId: Optional[str] = None
    providerCode: Optional[str] = None
    bypassCache: bool = False


class TranslateData(BaseModel):
    translatedContent: str
    sourceLanguage: Optional[str] = None
    targetLanguage: str
    model: Optional[str] = None
    tokensUsed: Optional[int] = None
    latencyMs: Optional[int] = None

File: apps/server-go/internal/dto/ai.go
package dto

// 说明：
//   Go 侧的 AI 请求 DTO 只做"声明性文档"用途——ai_handler 的业务端点通过
//   `proxySyncPost` / `proxySyncRequest` 直接把请求体原样转发给 Python FastAPI
//   微服务（见 internal/handler/ai_handler.go）。换句话说，Go 不 bind 这些
//   结构体，字段定义完全不进入网络协议层。
//
//   然而字段名仍需要和 Python Pydantic schema（apps/ai-service/app/schemas/ai.py）
//   保持严格一致，否则会误导前端开发者在客户端 TS 里面发一些 Python 根本
//   不识别的字段。本次清理移除了一批"幽灵字段"——Python 从未存在却长期
//   保留在 Go 侧的兼容别名：
//     - SummaryRequest:  Model / Style
//     - TagsRequest:     Model
//     - TitlesRequest:   Count / Style / Model
//     - PolishRequest:   PolishType / Style / Model
//     - OutlineRequest:  Model
//   保留 ModelID + ProviderCode（Python schema 中存在），供用户级模型覆盖使用。

// SummaryRequest 是 AI 摘要生成接口的请求 DTO。
type SummaryRequest struct {
	Content        string `json:"content" validate:"required"` // 需要生成摘要的正文内容（必填）
	MaxLength      int    `json:"maxLength,omitempty"`         // 摘要最大字数限制，默认 200，范围 10-2000
	PromptVersion  string `json:"promptVersion,omitempty"`     // 提示词版本号（可选）
	PromptTemplate string `json:"promptTemplate,omitempty"`    // 自定义提示词模板（可选）
	ModelID        string `json:"modelId,omitempty"`           // 模型 ID（可选，用户级覆盖）
	ProviderCode   string `json:"providerCode,omitempty"`      // AI 服务提供商代码（可选，用户级覆盖）
}

// TagsRequest 是 AI 标签推荐接口的请求 DTO。
type TagsRequest struct {
	Content        string `json:"content" validate:"required"` // 需要提取标签的正文内容（必填）
	MaxTags        int    `json:"maxTags,omitempty"`           // 最多返回的标签数量，默认 5，范围 1-20
	PromptVersion  string `json:"promptVersion,omitempty"`     // 提示词版本号（可选）
	PromptTemplate string `json:"promptTemplate,omitempty"`    // 自定义提示词模板（可选）
	ModelID        string `json:"modelId,omitempty"`           // 模型 ID（可选）
	ProviderCode   string `json:"providerCode,omitempty"`      // AI 服务提供商代码（可选）
}

// TitlesRequest 是 AI 标题推荐接口的请求 DTO。
type TitlesRequest struct {
	Content        string `json:"content" validate:"required"` // 需要生成标题的正文内容（必填）
	MaxTitles      int    `json:"maxTitles,omitempty"`         // 希望生成的标题数量，默认 5，范围 1-10
	PromptVersion  string `json:"promptVersion,omitempty"`     // 提示词版本号（可选）
	PromptTemplate string `json:"promptTemplate,omitempty"`    // 自定义提示词模板（可选）
	ModelID        string `json:"modelId,omitempty"`           // 模型 ID（可选）
	ProviderCode   string `json:"providerCode,omitempty"`      // AI 服务提供商代码（可选）
}

// PolishRequest 是 AI 内容润色接口的请求 DTO。
type PolishRequest struct {
	Content        string `json:"content" validate:"required"` // 需要润色的正文内容（必填）
	Tone           string `json:"tone,omitempty"`              // 润色语气（如"专业"/"学术"/"轻松"）
	PromptVersion  string `json:"promptVersion,omitempty"`     // 提示词版本号（可选）
	PromptTemplate string `json:"promptTemplate,omitempty"`    // 自定义提示词模板（可选）
	ModelID        string `json:"modelId,omitempty"`           // 模型 ID（可选）
	ProviderCode   string `json:"providerCode,omitempty"`      // AI 服务提供商代码（可选）
}

// OutlineRequest 是 AI 文章大纲生成接口的请求 DTO。
type OutlineRequest struct {
	Topic           string `json:"topic,omitempty"`           // 文章主题（topic 或 content 至少一个）
	Content         string `json:"content,omitempty"`         // 作为主题补充的正文内容（可选）
	ExistingContent string `json:"existingContent,omitempty"` // 已有的正文内容，供 AI 参考（可选）
	Depth           int    `json:"depth,omitempty"`           // 大纲层级深度，默认 2，范围 1-6
	Style           string `json:"style,omitempty"`           // 文章风格，默认 professional（professional / casual / technical）
	PromptVersion   string `json:"promptVersion,omitempty"`   // 提示词版本号（可选）
	PromptTemplate  string `json:"promptTemplate,omitempty"`  // 自定义提示词模板（可选）
	ModelID         string `json:"modelId,omitempty"`         // 模型 ID（可选）
	ProviderCode    string `json:"providerCode,omitempty"`    // AI 服务提供商代码（可选）
}

// TranslateRequest 是 AI 翻译接口的请求 DTO。
type TranslateRequest struct {
	Content        string `json:"content" validate:"required"` // 需要翻译的原文内容（必填）
	TargetLanguage string `json:"targetLanguage,omitempty"`    // 目标语言（可选，如 zh/en/ja）
	SourceLanguage string `json:"sourceLanguage,omitempty"`    // 源语言（可选，不填则自动检测）
	PromptVersion  string `json:"promptVersion,omitempty"`     // 提示词版本号（可选）
	PromptTemplate string `json:"promptTemplate,omitempty"`    // 自定义提示词模板（可选）
	ModelID        string `json:"modelId,omitempty"`           // 模型 ID（可选）
	ProviderCode   string `json:"providerCode,omitempty"`      // AI 服务提供商代码（可选）
}

// PromptUpdateRequest 是更新 AI 提示词模板的请求 DTO。
// 注：JSON tag 使用 snake_case（`prompt_template`），与 Python Pydantic schema
// （apps/ai-service/app/api/routes/prompts.py::PromptUpdateRequest）保持一致。
type PromptUpdateRequest struct {
	PromptTemplate string `json:"prompt_template"` // 新的提示词模板内容
}

// TaskTypeCreateRequest 是创建 AI 任务类型的请求 DTO。
// 注：字段使用 snake_case JSON tag，与 Python Pydantic schema 一致。
type TaskTypeCreateRequest struct {
	Code           string   `json:"code"`                      // 任务类型唯一代码标识
	Name           string   `json:"name"`                      // 任务类型显示名称
	Description    string   `json:"description,omitempty"`     // 任务类型描述（可选）
	ModelType      string   `json:"model_type,omitempty"`      // 适用的模型类型（可选）
	Temperature    *float64 `json:"temperature,omitempty"`     // 生成温度，控制随机性，范围 0~2（可选）
	MaxTokens      *int     `json:"max_tokens,omitempty"`      // 最大生成 token 数量（可选）
	PromptTemplate string   `json:"prompt_template,omitempty"` // 提示词模板内容（可选）
}

// TaskTypeUpdateRequest 是更新 AI 任务类型的请求 DTO，所有字段均为可选。
type TaskTypeUpdateRequest struct {
	Code           string   `json:"code,omitempty"`            // 任务类型唯一代码标识（可选）
	Name           string   `json:"name,omitempty"`            // 任务类型显示名称（可选）
	Description    string   `json:"description,omitempty"`     // 任务类型描述（可选）
	ModelType      string   `json:"model_type,omitempty"`      // 适用的模型类型（可选）
	Temperature    *float64 `json:"temperature,omitempty"`     // 生成温度，控制随机性，范围 0~2（可选）
	MaxTokens      *int     `json:"max_tokens,omitempty"`      // 最大生成 token 数量（可选）
	PromptTemplate string   `json:"prompt_template,omitempty"` // 提示词模板内容（可选）
}
